mitiq/zne.py

```diff
@@ -1,147 +1,50 @@
-"""Zero noise extrapolation sub-module."""
+from typing import List, Tuple, Callable
+import mitiq.qiskit.qiskit_utils as qs_utils
+from mitiq import QPROGRAM
+from mitiq.factories import Factory, LinearFactory
 
-import numpy as np
 
-
-# used in Mitigator.extrapolate()
-def get_gammas(c):
-    """Returns the linear combination coefficients "gammas" for Richardson's extrapolation.
-    The input is a list of the noise stretch factors.
+def mitigate(pq: QPROGRAM, fac: Factory, scale_noise: Callable[[QPROGRAM, float], QPROGRAM], run_program: Callable) -> \
```

Comment:
I would rename this function. `mitigate` should conceptually be `zne` below: i.e., a function take in 

1. a circuit,
2.  fold function,
3. a stretch factor, and 
4. an extrapolation method, 

then return the extrapolated zero noise (i.e., mitigated) value.

Additionally, it seems like this should be a method of a `Factory`, not a standalone function. It's odd to input a `Factory`.

Reply:
Actually, I just realized `zne` returns a function. I admit I'm confused by this as well.

Reply:
The reason to input a factory is to decouple the backend that runs programs from the method of doing ZNE. You can define one Factory for zne and then run it on a bunch of different backends with mitigate. If they were coupled then you'd need to define a Factory with a backend specified at instantiation. Conceptually I think it is nice to assume that Factory objects are totally backend independent.

`zne` returns a function so that it can be a drop in replacement for wherever a user has `run_program` in their code, e.g.

```
qvm = QVMConnection()

output = qvm.run(my_pyquil_prog)

# becomes
zne_run = zne(qvm.run) # with default args to LinearFactory
output = zne_run(my_pyquil_program)
```